gax/backoff_policy.h

```diff
@@ -0,0 +1,181 @@
+  virtual std::chrono::microseconds MaximumDelay() const = 0;
+};
+
+/**
+ * Implements a truncated exponential backoff with randomization policy.
+ *
+ * This policy implements the truncated exponential backoff policy for
+ * retrying operations. After a request fails, and subject to a separate
+ * retry policy, the client library will wait for an initial delay before
+ * trying again. If the second attempt fails the delay time is increased by a factor
+ * of 2. The delay time growth stops at a maximum delay wait time.
+ * The policy also randomizes the delay each time, to avoid
+ * [thundering herd
+ * problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
+ *
+ * Note: The random number generator used when calculating backoff time in
+ *       OnCompletion is lazily created. Furthermore, GeneratorFactory::Generator
+ *       may touch mutable shared state.
+ *       As a result, OnCompletion is NOT thread-safe.
+ *       All other methods, including clone, are thread-safe.
+ *       ExponentialBackoffPolicy is therefore thread-compatible.
+ */
+template <typename GeneratorFactory>
+    class ExponentialBackoffPolicy : BackoffPolicy {
+ public:
+  /**
+   * Constructor for an exponential backoff policy.
+   *
+   * Define the initial delay and maximum delay for an instance
+   * of the policy. While the constructor accepts `std::chrono::duration`
+   * objects at any resolution, the data is kept internally in microseconds.
+   * Sub-microsecond delays seem unnecessarily precise for this application.
+   *
+   * @code
+   * using namespace std::chrono_literals; // C++14
+   * auto r1 = ExponentialBackoffPolicy(10ms, 500ms);
+   * @endcode
+   *
+   * @param initial_delay how long to wait after the first (unsuccessful)
+   *     operation.
+   * @param maximum_delay the maximum value for the delay between operations.
+   *
+   * @tparam duration1_t a placeholder to match the Rep tparam for @p initial_delay's
+   *     type, the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+   *     used to store the number of ticks), for our purposes it is simply a
+   *     formal parameter.
+   * @tparam d1 a placeholder to match the Period tparam for
+   *     @p initial_delay's type, the semantics of this template parameter are
+   *     documented in `std::chrono::duration<>` (in brief, the length of the
+   *     tick in seconds, expressed as a `std::ratio<>`), for our purposes it
+   *     is simply a formal parameter.
+   * @tparam duration2_t similar formal parameter for the type of @p maximum_delay.
+   * @tparam d2 similar formal parameter for the type of @p maximum_delay.
+   *
+   * @see
+   * [std::chrono::duration<>](http://en.cppreference.com/w/cpp/chrono/duration)
+   *     for more details.
+   */
+  template<typename duration1_t, typename duration2_t>
+      ExponentialBackoffPolicy(duration1_t d1, duration2_t d2) :
```

Comment:
Nit: indentation

Reply:
Emacs c++ mode keeps wigging out on me. I'll take a look under the hood and see if I can fix it permanently.